### PR TITLE
reverse ens lookup for addresses

### LINKED ISSUE #159: add ENS support
check accounts for ENS names and display if they exist
gitcoin bounty possible? @Arachnid @decanus

### LINKED ISSUE #169: "delete clover" (sell back to bank?)
going into My Account, realize I'd like to just remove some dregs. how might we handle that?

### PR DESCRIPTION
**Pull request by StevenJNPearce**
Reverse ens lookup for addresses

Fixes #159

### CODE REVIEW

**Comment by okwme** [comment]
Hi thanks very much for the PR!

i'm not sure that async computed is the best solution for this though. This will trigger the ens call every time `owner` is computed which includes every time `allUsers` changes. It will also have redundant calls when the same page is seen multiple times on the same visit.

Maybe it's better to add `ens` as a property to the user in the `allUsers` array in the vuex state which can be checked inside of `addrToUser`.

in that case there would be a vuex action like "updateUserENS" that is triggered when `user.ens` is `undefined` inside of `addrToUser`. After the action pings ens it would trigger a vuex mutation that updates that user with the new key `ens` set to be the ens name or `false` to avoid pinging them a second time.

File: src/utils.js [hunk 1]
```diff
@@ -9,13 +9,23 @@ export function pad0x (string) {
   return `0x${string}`
 }
 
-export function addrToUser (allUsers, address) {
+export async function addrToUser (allUsers, address) {
   let userIndex = allUsers.findIndex(
     u => u.address.toLowerCase() === address.toLowerCase()
   )
+  address = await getEnsName(address)
   return userIndex > -1 ? allUsers[userIndex].name || address : address
```
Comment (okwme) [inline src/utils.js]:
```
  return userIndex > -1 && allUsers[userIndex].name || await getEnsName(address)
```

**Comment by okwme** [comment]
HI @StevenJNPearce 
good question: no, it is not possible for a user to own a clover and not appear in the allUsers list. This list is populated by an event listener on our api for any contract activity and adds new users when they appear. So no need to add them to allUsers if they don't exist.

maybe something like this makes sense:
```
return userIndex > -1 && allUsers[userIndex].name || (() => {
  if ( allUsers[userIndex].ensChecked) {
    return allUsers[userIndex].ens || address
  } else {
    dispatch('getEns', address)
    return address
  }
})()
```
where `getEns()` checks the address and then mutates the `allUsers` array to update that user's `ensChecked` and `ens` attributes?

**Comment by StevenJNPearce** [comment]
Ok, I have implemented the changes and rebased against upstream.

File: src/store/index.js [hunk 2]
```diff
@@ -8,9 +8,12 @@ import mutations from './mutations'
 
 import Reversi from 'clovers-reversi'
 import BigNumber from 'bignumber.js'
+import AsyncComputed from 'vue-async-computed'
```
Comment (okwme) [inline src/store/index.js]: no longer needed now, right?

File: src/utils.js [hunk 3]
```diff
@@ -13,7 +14,23 @@ export function addrToUser (allUsers, address) {
...
+    if (allUsers[userIndex].ens === undefined) {
+      store.dispatch('updateUserENS', allUsers[userIndex])
+    } else if (allUsers[userIndex].ens !== false) {
+      return allUsers[userIndex].ens
+    }
+    return address
+  })()
+}
+
+export async function getEnsName (address) {
```
Comment (okwme) [inline src/utils.js]: also no longer needed, right?

File: package.json [hunk 4]
```diff
@@ -49,6 +50,7 @@
     "truffle-contract": "^2.0.5",
     "underscore": "^1.9.1",
     "vue": "^2.5.13",
+    "vue-async-computed": "^3.4.1",
```
Comment (okwme) [inline package.json]: not needed anymore?

**Comment by okwme** [comment]
looks great @StevenJNPearce thanks
if we can remove the stuff that's not needed anymore i'll be happy to merge : )

**Comment by StevenJNPearce** [comment]
@okwme  Done :)

**Comment by vs77bb** [comment]
Hi @okwme is this good for merge / payout?

**Comment by StevenJNPearce** [comment]
@okwme Your output on github is prolific, I'm guessing this got lost in a string of notifications, so I'm pinging you now as a reminder. See @vs77bb's comment above.

**Comment by okwme** [comment]
hey @StevenJNPearce & @vs77bb 

sorry for the delay, I've been in the process of moving the last couple weeks and am only now getting settled and able to focus on clovers full time. 

I've merged the work into the dev branch, it actually doesn't work since the clovers is only on rinkeby for now, but I can either add a custom provider for readonly on mainnet, or just wait until we're actually on mainnet to confirm that it works. I guess this bounty should be closed anyway tho, thank you for your work Steven!

**Pull request merged**